New GitHub action workflow for pages (#1165)
withastro/docs: src/pages/en/guides/deploy/github.md

2. Create a new file in your project at `.github/workflows/deploy.yml` and paste in the YAML below.

    ```yaml

>     name: Github Pages Astro CI

Reviewer: Would also leave the old name

    name: Build astro site

    on:

>       # Trigger the workflow every time you push to the `main` branch

Reviewer: Comments missing again

      # Using a different branch name? Replace `main` with your branch’s name
      push:
        branches: [main]
      # Allows you to run this workflow manually from the Actions tab on GitHub.

>       workflow_dispatch:

Reviewer: Is this still possible or not?

Reviewer: (Meaning workflow_dispatch or is it not a real action anymore?)

Author: still possible! sorry I should've double checked how the old action looked. I kinda just replaced it with the workflow my team made!

        branches: [ main ]
      pull_request:
        branches: [ main ]

    permissions:
      contents: read
      pages: write
      id-token: write

    jobs:
      deploy:
        runs-on: ubuntu-20.04
      build:

        # Allow this job to push changes to your repository
        permissions:
          contents: write
        runs-on: ubuntu-latest

        steps:
          - name: Check out your repository using git
            uses: actions/checkout@v2

          - name: Use Node.js 16
            uses: actions/setup-node@v2
            with:
              node-version: 16

>           # Not using npm? Change `npm ci` to `yarn install` or `pnpm i`

Reviewer: I would leave the comment there

Author: okay adding it in!!!

Reviewer: Thx

          - name: Install dependencies
            run: npm ci

>           # Not using npm? Change `npm run build` to `yarn build` or `pnpm run build`

Reviewer: Same applies to here

    name: Github Pages Astro CI

    on:

>      # Trigger the workflow every time you push to the `main` branch

Reviewer 2: Do you think it would be helpful to include this notice we used to have for folks have have different branch names?

```suggestion
      # Trigger the workflow every time you push to the `main` branch
      # Using a different branch name? Replace `main` with your branch’s name
```

Reviewer 3: @delucis @blackgirlbytes If this instruction is still correct (replace `main` if your'e using a different branch), then I'd be in favour of keeping it in. I would say, only remove if it's wrong or misleading.

      push:
        branches: [ main ]
      # Allows you to run this workflow manually from the Actions tab on GitHub.
      workflow_dispatch:
      

>       # Allow this job to push changes to your repository

Reviewer 2: Nit: looks like this needs slightly different indentation to align with the block it refers to:

<img width="190" alt="image" src="https://user-images.githubusercontent.com/357379/182553128-a4d24844-d250-4606-b5a8-59a6ba8a33e8.png">

```suggestion
    # Allow this job to push changes to your repository
```

    permissions:
      contents: read
      pages: write
      id-token: write

    jobs:
      build:
        runs-on: ubuntu-latest
        steps:

>         - uses: actions/checkout@v2
>         - uses: actions/setup-node@v2
>           with:
>             node-version: '16'
>             cache: 'npm'
>         # Not using npm? Change `npm ci` to `yarn install` or `pnpm i`
>         - run: npm ci
>         # Not using npm? Change `npm run build` to `yarn build` or `pnpm run build`
>         - run: npm run build --if-present

Reviewer 2: What do you think about including a `name` for each of these steps?

It is more verbose, but I kind of like how it documents what each thing does — especially for useful for those less familiar with GitHub Actions.

```suggestion
        - name: Check out your repository using git
          uses: actions/checkout@v2

        - name: Use Node.js 16
          uses: actions/setup-node@v2
          with:
            node-version: '16'
            cache: 'npm'

        # Not using npm? Change `npm ci` to `yarn install` or `pnpm i`
        - name: Install dependencies
          run: npm ci

        # Not using npm? Change `npm run build` to `yarn build` or `pnpm run build`
        - name: Build Astro
          run: npm run build --if-present
```

Author: yeah I like that. Very clear for beginners!

        - name: Archive build output
          # `./dist` is the default Astro build directory.
          # If you changed that, update it here too.
          run: "tar --dereference --directory dist/ -cvf artifact.tar ."

>         - name: Upload artifact

Reviewer 2: Nit: for consistency I guess we should have a new line before this step too.

```suggestion

        - name: Upload artifact
```

    name: Github Pages Astro CI

    on:
     # Trigger the workflow every time you push to the `main` branch
      push:
        branches: [ main ]
      # Allows you to run this workflow manually from the Actions tab on GitHub.
      workflow_dispatch:
      

>       # Allow this job to clone the repo and create a page deployment

Reviewer 2: ```suggestion
    # Allow this job to clone the repo and create a page deployment
```

    permissions:
      contents: read
      pages: write
      id-token: write

    jobs:
      build:
        runs-on: ubuntu-latest
        steps:
        - name: Check out your repository using git
          uses: actions/checkout@v2

        - name: Use Node.js 16
          uses: actions/setup-node@v2
          with:
            node-version: '16'
            cache: 'npm'

        # Not using npm? Change `npm ci` to `yarn install` or `pnpm i`
        - name: Install dependencies
          run: npm ci

        # Not using npm? Change `npm run build` to `yarn build` or `pnpm run build`
        - name: Build Astro

>           run: npm run build --if-present

Reviewer 4: I'm not an Astro expert but, from what I've seen in the docs, the build is required for publishing to GitHub Pages. As such, I would leave the `--if-present` off since that effectively makes this step optional.

```suggestion
          run: npm run build
```

Author: good point! thank you!

    name: Github Pages Astro CI

    on:
     # Trigger the workflow every time you push to the `main` branch
      push:
        branches: [ main ]
      # Allows you to run this workflow manually from the Actions tab on GitHub.
      workflow_dispatch:
      

>       # Allow this job to push changes to your repository

Reviewer 5: Minor rephrasing. Part of the update to the new workflow is we no longer require a token that can push to the repo directly:

```suggestion
      # Allow this job to clone the repo and create a page deployment
```

    permissions:
      contents: read
      pages: write
      id-token: write

    jobs:
      build:
        runs-on: ubuntu-latest
        steps:
        - name: Check out your repository using git
          uses: actions/checkout@v2

        - name: Use Node.js 16
          uses: actions/setup-node@v2
          with:
            node-version: '16'
            cache: 'npm'

        # Not using npm? Change `npm ci` to `yarn install` or `pnpm i`
        - name: Install dependencies
          run: npm ci

        # Not using npm? Change `npm run build` to `yarn build` or `pnpm run build`
        - name: Build Astro
          run: npm run build --if-present

>         - name: Archive build output
>           # `./dist` is the default Astro build directory.
>           # If you changed that, update it here too.
>           run: "tar --dereference --directory dist/ -cvf artifact.tar ."
> 
>         - name: Upload artifact
>           uses: actions/upload-artifact@v2
>           with:
>             name: github-pages
>             path: artifact.tar

Reviewer 5: These steps have a composite action. See example [here](https://github.com/actions/starter-workflows/blob/4a8f18e34dd13d2b6ee4d8da2ba72629eafe1609/pages/nextjs.yml#L80-L83)

```suggestion
      - name: Upload artifact
        uses: actions/upload-pages-artifact@v1
        with:
          path: ./dist
```

      deploy:
        environment:
          name: github-pages
          url: ${{ steps.deployment.outputs.page_url }}
        runs-on: ubuntu-latest
        needs: build
        steps:
          - name: Deploy to GitHub Pages
            id: deployment

>             uses: paper-spa/deploy-pages@main

Reviewer 5: This is an org we use for staging and could break since it's from `@main`. Better to use the released actions from the actions org

```suggestion
            uses: actions/deploy-pages@v1
```

      # Trigger the workflow every time you push to the `main` branch
      # Using a different branch name? Replace `main` with your branch’s name
      push:
        branches: [ main ]
      # Allows you to run this workflow manually from the Actions tab on GitHub.
      workflow_dispatch:
      
    # Allow this job to clone the repo and create a page deployment
    permissions:
      contents: read
      pages: write
      id-token: write

    jobs:
      build:
        runs-on: ubuntu-latest
        steps:
        - name: Check out your repository using git
          uses: actions/checkout@v2

        - name: Use Node.js 16
          uses: actions/setup-node@v2
          with:
            node-version: '16'
            cache: 'npm'

        # Not using npm? Change `npm ci` to `yarn install` or `pnpm i`
        - name: Install dependencies
          run: npm ci

        # Not using npm? Change `npm run build` to `yarn build` or `pnpm run build`
        - name: Build Astro
          run: npm run build

>       - name: Upload artifact
>         uses: actions/upload-pages-artifact@v1
>         with:
>           path: ./dist

Reviewer 2: Oh, looks like I missed this indentation difference here — not sure it’s critical, but updating to align with the steps above just to be safe.

```suggestion
        - name: Upload artifact
          uses: actions/upload-pages-artifact@v1
          with:
            path: ./dist
```
